lib/div.cpp

```diff
@@ -0,0 +1,457 @@
+#include "flag_gems/operators.h"
+#include "flag_gems/utils.h"
+
+#include <c10/cuda/CUDAStream.h>
+#include <c10/util/TypeTraits.h>
+#include <iostream>
+#include "triton_jit/triton_jit_function.h"
+
+namespace flag_gems {
+using namespace triton_jit;
+
+at::Tensor true_div(const at::Tensor& a_, const at::Tensor& b_) {
+  // Case 1: Scalar / Scalar → fallback
+  if (a_.dim() == 0 && b_.dim() == 0) {
+    double a_val = a_.item<double>();
+    double b_val = b_.item<double>();
+    return torch::tensor(a_val / b_val, a_.options());
+  }
+
+  // Case 2: Scalar / Tensor
+  if (a_.dim() == 0 && b_.dim() > 0) {
+    at::Tensor a = at::full_like(b_, a_.item<double>());
+    return true_div(a, b_);
+  }
+
+  // Case 3: Tensor / Scalar
+  if (a_.dim() > 0 && b_.dim() == 0) {
+    at::Tensor b = at::full_like(a_, b_.item<double>());
+    return true_div(a_, b);
+  }
+
+  // Case 4: Tensor / Tensor
+  auto res = torch::broadcast_tensors({a_, b_});
+  at::Tensor a = res[0].contiguous();
+  at::Tensor b = res[1].contiguous();
+
+  at::ScalarType out_dtype = at::promote_types(a.scalar_type(), b.scalar_type());
+  at::Tensor out = at::empty(a.sizes(), at::TensorOptions().dtype(out_dtype).device(a.device()));
+
+  const TritonJITFunction& f =
+      TritonJITFunction::getInstance((utils::get_triton_src_path() / "div.py").string(), "true_div_kernel");
+
+  int64_t tile_size = 1024;
+  int64_t n = out.numel();
+  int num_blocks = (n + tile_size - 1) / tile_size;
+
+  c10::cuda::CUDAStream stream = c10::cuda::getCurrentCUDAStream();
+  CUstream raw_stream = static_cast<CUstream>(stream.stream());
+  c10::DeviceGuard guard(out.device());
+
+  f(raw_stream, num_blocks, 1, 4, 2, 0, a, b, out, n, tile_size);
+
+  return out;
+}
+
+at::Tensor true_div_(at::Tensor& a_, const at::Tensor& b_) {
+  TORCH_CHECK(a_.device().is_cuda(), "true_div_: only CUDA tensors supported");
+  TORCH_CHECK(a_.is_contiguous(), "true_div_: input a_ must be contiguous for in-place op");
+
+  //  Tensor // Scalar
+  if (a_.dim() > 0 && b_.dim() == 0) {
+    at::Tensor b_expand = at::full_like(a_, b_.item<double>());
+    return true_div_(a_, b_expand);
+  }
+
+  // Tensor // Tensor
+  auto res = torch::broadcast_tensors({a_, b_});
+  at::Tensor a_broadcast = res[0].contiguous();
+  at::Tensor b_broadcast = res[1].contiguous();
+  TORCH_CHECK(a_.numel() == a_broadcast.numel(), "true_div_: broadcasted shape mismatch");
+
+  TORCH_CHECK(a_broadcast.dim() >= 2, "true_div_: input must be at least 2D");
+  int64_t M = a_broadcast.size(0);
+  int64_t N = a_broadcast.numel() / M;
+
+  at::Tensor out = at::empty_like(a_broadcast);
+
+  const TritonJITFunction& f =
+      TritonJITFunction::getInstance((utils::get_triton_src_path() / "div.py").string(), "true_div_kernel_");
+
+  int64_t tile_size = 1024;
+  int64_t n = M * N;
+  int num_blocks = (n + tile_size - 1) / tile_size;
```

Comment:
I'm not sure why the input is viewed as a 2d tensor. Once you've made the inputs contiguous, its simply a 1d elementwise operation.